File: Binary Search/K-diff Pairs in an Array.cpp
// TC:O(n^2) SC:O(1)
class Solution {
public:
    int findPairs(vector<int>& nums, int k) {
        sort(nums.begin(), nums.end());
        int maxCnt = 0;
        for (int i = 0; i < nums.size(); i++) {
            if (i != 0 && nums[i] == nums[i - 1])
                continue;
            for (int j = i + 1; j < nums.size(); j++) {

                if (abs(nums[i] - nums[j]) == k) {
                    maxCnt++;
                    break;
                } else if (abs(nums[i] - nums[j]) > k) {
                    break;
                }
            }
        }
        return maxCnt;
    }
};

File: Binary Search/Python/K-diff Pairs in an Array.py
// TC:O(n^2) SC:O(1)
class Solution:
    def findPairs(self, nums: List[int], k: int) -> int:
        nums.sort()
        maxCnt = 0
        n = len(nums)

        for i in range(n):
            if i > 0 and nums[i] == nums[i - 1]:
                continue

            for j in range(i + 1, n):
                if abs(nums[i] - nums[j]) == k:
                    maxCnt += 1
                    break
                elif abs(nums[i] - nums[j]) > k:
                    break
        return maxCnt

//TC:O(n log n) SC:O(1)
class Solution:
    def findPairs(self, nums: List[int], k: int) -> int:
        nums.sort()
        maxCnt=0
        n = len(nums)
        for i in range(n):
            if i!=0 and nums[i]==nums[i-1]:continue
            comp=nums[i]+k
            low=i+1
            high=n-1
            while low<=high:
                mid=low+(high-low)//2
                if(nums[mid]==comp):
                    maxCnt=maxCnt+1
                    break
                elif nums[mid]>comp:
                    high=mid-1
                else: low=mid+1
        return maxCnt

        

        
